src/core/manager.c

```diff
@@ -1828,17 +1829,19 @@ static int manager_dispatch_signal_fd(sd_event_source *source, int fd, uint32_t
         }
 
         for (;;) {
-                n = read(m->signal_fd, &sfsi, sizeof(sfsi));
-                if (n != sizeof(sfsi)) {
-
-                        if (n >= 0)
-                                return -EIO;
-
-                        if (errno == EINTR || errno == EAGAIN)
+                ttl = 0;
+                do {
+                        n = read(m->signal_fd, buf + ttl, length - ttl);
+                        if (n < 0 && errno == EAGAIN) {
                                 break;
-
-                        return -errno;
-                }
+                        } else if (n < 0 && errno != EINTR) {
+                                log_error_errno(errno, "Failed to read bytes from signal file descriptor (%d): %m", errno); return -errno;
+                        } else if (n == 0) {
+                            log_error("Signal file descriptor received client disconnect, with %ld bytes discarded.", ttl); return -EIO; }
```

Comment:
The signalfd(2) call causes signals to be delivered over a socket, instead of using other methods known to cause problems with multithreaded applications. Please see the relevant man pages and examples on the web for additional information.

Comment #3634 does not look relevant to this pull request, but rather to rate limiting them. This pull request is about properly reading them to pass to #3634 for processing...